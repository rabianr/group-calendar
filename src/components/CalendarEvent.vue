<script setup lang="ts">
import { ref } from 'vue'
import { format } from 'date-fns'
import { Popover } from 'bootstrap'
import type { DisplayEvent } from './CalendarEvent.types'
import CalendarEventDetail from './CalendarEventDetail.vue'
import CalendarCellSpacer from './CalendarCellSpacer.vue'

const props = defineProps<{
  data: DisplayEvent
}>()

const calendarEventDetailRef = ref()
const hasPopover = ref(false)

const showDetail = (e: Event) => {
  e.preventDefault()

  const target = e.target as HTMLElement

  // イベント（aタグ）がクリックされない場合は処理を中止
  if (!target.classList.contains('event')) {
    return
  }

  const popover = new Popover(target, {
    title: props.data.title,
    content: calendarEventDetailRef.value.tableRef,
    customClass: 'popover-event-detail',
    container: target,
    html: true,
    placement: 'bottom',
    trigger: 'focus'
  })

  target.addEventListener(
    'hidden.bs.popover',
    () => {
      popover.dispose()
      hasPopover.value = false
    },
    { once: true }
  )

  hasPopover.value = true
  popover.show()
}
</script>

<template>
  <a
    href="#"
    class="event"
    :class="{ 'has-popover': hasPopover }"
    :style="{ '--colspan': data.colspan, '--pos': data.pos }"
    @click="showDetail"
    tabindex="0"
  >
    <template v-if="data.start != data.end">
      {{ format(data.start, 'MM/dd') }} - {{ format(data.end, 'MM/dd') }}
    </template>
    {{ data.title }}
  </a>
  <div class="d-none">
    <CalendarEventDetail ref="calendarEventDetailRef" :data="data" />
  </div>
  <CalendarCellSpacer v-if="data.last" :lastPos="data.pos + 1" />
</template>

<style lang="scss" scoped>
.event {
  --pos: 0;
  --colspan: 1;
  position: absolute;
  top: calc(var(--pos) * (var(--evt-height) + var(--evt-margin)));
  left: var(--evt-margin);
  z-index: 10;
  width: calc(
    100% * var(--colspan) - var(--evt-margin, 0) - var(--evt-margin, 0) +
      ((var(--colspan) - 1) * 1px)
  );
  height: var(--evt-height);
  display: block;
  background: var(--bs-primary);
  color: var(--bs-light);
  font-size: 0.85rem;
  text-decoration: none;
  margin-top: var(--evt-margin);
  padding: 1px 3px;
  border-radius: 3px;

  &.has-popover {
    z-index: 20;
  }
}
</style>

<style lang="scss">
.popover-event-detail {
  --bs-popover-max-width: 400px;
  --bs-popover-border-color: var(--bs-dark);
  --bs-popover-header-bg: var(--bs-primary);
  --bs-popover-header-color: var(--bs-white);
  --bs-popover-body-padding-x: 1rem;
  --bs-popover-body-padding-y: 0.5rem;
  width: var(--bs-popover-max-width);
  box-shadow: var(--bs-popover-box-shadow);

  .popover-arrow {
    --bs-popover-bg: var(--bs-primary);
  }

  .popover-header {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
}
</style>
