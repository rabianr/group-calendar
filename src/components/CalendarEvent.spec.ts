import { describe, expect, it, vi } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import { nextTick } from 'vue'
import CalendarEvent from './CalendarEvent.vue'

describe('CalendarEvent コンポーネント', () => {
  const data = {
    title: 'テスト',
    detail: 'テスト内容',
    start: '2024-04-10',
    end: '2024-04-10',
    colspan: 1,
    pos: 0,
    last: false
  }

  it('スナップショットテスト', async () => {
    const wrapper = shallowMount(CalendarEvent, {
      global: {
        plugins: [createTestingPinia({ createSpy: vi.fn })]
      },
      props: {
        data
      }
    })

    const eventWrapper = wrapper.find('.event')
    expect(eventWrapper.text()).toContain('テスト')
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ data: { ...data, end: '2024-04-11', colspan: 2, pos: 1, last: true } })
    expect(eventWrapper.text()).toContain('04/10 - 04/11 テスト')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('イベント詳細を描画する', async () => {
    const wrapper = mount(CalendarEvent, {
      attachTo: document.body,
      global: {
        plugins: [createTestingPinia({ createSpy: vi.fn })]
      },
      props: {
        data
      }
    })

    const eventWrapper = wrapper.find('.event')

    // Popoverを表示
    await eventWrapper.trigger('click')
    expect(eventWrapper.classes()).toContain('has-popover')
    expect(eventWrapper.find('.popover').exists()).toBeTruthy()

    // Popoverを非表示
    eventWrapper.element.dispatchEvent(new Event('hidden.bs.popover'))
    await nextTick()
    expect(eventWrapper.classes()).not.toContain('has-popover')
    expect(eventWrapper.find('.popover').exists()).toBeFalsy()
  })
})
