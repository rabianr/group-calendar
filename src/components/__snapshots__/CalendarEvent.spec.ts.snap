// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CalendarEvent コンポーネント > スナップショットテスト 1`] = `
"<a data-v-bbbf80c4="" href="#" class="event" style="--colspan: 1; --pos: 0;" tabindex="0">
  <!--v-if--> テスト
</a>
<div data-v-bbbf80c4="" class="d-none">
  <calendar-event-detail-stub data-v-bbbf80c4="" data="[object Object]"></calendar-event-detail-stub>
</div>
<!--v-if-->"
`;

exports[`CalendarEvent コンポーネント > スナップショットテスト 2`] = `
"<a data-v-bbbf80c4="" href="#" class="event" style="--colspan: 2; --pos: 1; --pos: 1;" tabindex="0">04/10 - 04/11 テスト</a>
<div data-v-bbbf80c4="" class="d-none">
  <calendar-event-detail-stub data-v-bbbf80c4="" data="[object Object]"></calendar-event-detail-stub>
</div>
<calendar-cell-spacer-stub data-v-bbbf80c4="" lastpos="2"></calendar-cell-spacer-stub>"
`;
